### ISSUE TITLE
Question: NativeAOT-LLVM: How do CSE vars get their definition

### ISSUE THREAD

**Issue by yowl** [dotnet/runtimelab#2036]
Looking at the attached log
[1.txt](https://github.com/dotnet/runtimelab/files/9781478/1.txt)

There is a CSE var `V02` and after the LIR phase its definition is after the use, reading the blocks in the order they are printed, not the flow order
```
------------ BB01 [000..005) -> BB03 (always), preds={} succs={BB03}
               [000039] ------------                 IL_OFFSET void   IL offset: 0x0
N001 (  1,  1) [000000] ------------         t0 =    CNS_INT   int    0 $40
                                                  /--*  t0     int
N003 (  1,  3) [000002] DA----------              *  STORE_LCL_VAR int    V00 loc0         d:2

------------ BB02 [005..00E), preds={BB04} succs={BB03}
N001 (  1,  1) [000033] ------------        t33 =    LCL_VAR   int    V02 cse0         u:1 <l:$141, c:$c1>
                                                  /--*  t33    int
N003 (  1,  3) [000018] DA--G-------              *  STORE_LCL_VAR int    V01 loc1         d:2
N001 (  1,  1) [000019] ------------        t19 =    LCL_VAR   int    V01 loc1         u:2 (last use) <l:$141, c:$c4>
                                                  /--*  t19    int
N003 (  1,  3) [000021] DA----------              *  STORE_LCL_VAR int    V00 loc0         d:4

------------ BB03 [00E..018) -> BB05 (cond), preds={BB01,BB02} succs={BB04,BB05}
N001 (  1,  1) [000003] ------------         t3 =    LCL_VAR   int    V00 loc0         u:3 $140
N002 (  1,  1) [000024] ------------        t24 =    CNS_INT   int    8 field offset Fseq[next] $44
                                                  /--*  t3     int
                                                  +--*  t24    int
N003 (  2,  2) [000025] -------N----        t25 = *  ADD       int    $101
                                                  /--*  t25    int
N004 (  4,  4) [000004] ---XG-------         t4 = *  IND       int    <l:$103, c:$102>
                                                  /--*  t4     int
N006 (  4,  4) [000035] DA-XG-------              *  STORE_LCL_VAR int    V02 cse0         d:1
N007 (  1,  1) [000036] ------------        t36 =    LCL_VAR   int    V02 cse0         u:1 <l:$103, c:$102>
```
The `localVar` function in the LLVM backend doesn't like this currently as it wants to find the defintiion, and the LLVM `Value*`, when it encounters the use:

https://github.com/dotnet/runtimelab/blob/32f5c33e5ad714fe4464e2f7937bd76b13a4c7a9/src/coreclr/jit/llvm.cpp#L1848-L1865

What would be a smart solution for this, use a dummy/null LLVM `Value*`, record in a map and replace as the definitions are encountered (I think that should be possible)? Process the blocks in flow order?  Grateful for any suggestions.

cc @SingleAccretion.

**Comment by SingleAccretion** [comment]
> Process the blocks in flow order?

It does seem the simplest option, a pre-order walk of the dominator tree SSA constructs. One potential downside is losing some block reordering benefits from PGO, but I'd probably actually trust LLVM more in that (and we should still have the block weighs around by then, even if a bit degraded).

**Comment by yowl** [comment]
Something like https://github.com/dotnet/runtimelab/blob/3eea0059430f2f1bead66c33c21f49d5bd9accc8/src/coreclr/jit/ssabuilder.cpp#L160 but pre order.  I.e. I can use a `ArrayStack<AllSuccessorEnumerator>` , `TopRef()`, `NextSuccessor` ?

**Comment by SingleAccretion** [comment]
More like this: https://github.com/dotnet/runtimelab/blob/3eea0059430f2f1bead66c33c21f49d5bd9accc8/src/coreclr/jit/copyprop.cpp#L467-L489.

**Comment by yowl** [comment]
Thanks, this should be closed by #2042 but just want to get the optimisation on so I can test with the sample that is in the log at the top of this page.
